pkg/skaffold/deploy/labels.go

```diff
@@ -269,46 +116,81 @@ func labelDeployResults(labels map[string]string, results []Artifact) {
-		m.Labels[k] = v
+	for key, value := range labels {
+		objLabels[key] = value
 	}
+	accessor.SetLabels(objLabels)
 }
 
-func retrieveNamespace(ns string, m metav1.ObjectMeta) string {
+func updateRuntimeObject(client dynamic.Interface, disco discovery.DiscoveryInterface, labels map[string]string, res Artifact) error {
+	originalJSON, _ := json.Marshal(*res.Obj)
+	modifiedObj := (*res.Obj).DeepCopyObject()
+	accessor, err := meta.Accessor(modifiedObj)
+	if err != nil {
+		return errors.Wrap(err, "getting metadata accessor")
+	}
+	name := accessor.GetName()
+	namespace := accessor.GetNamespace()
+	addLabels(labels, accessor)
+
+	modifiedJSON, _ := json.Marshal(modifiedObj)
+	p, _ := patch.CreateTwoWayMergePatch(originalJSON, modifiedJSON, modifiedObj)
+	gvr, err := groupVersionResource(disco, modifiedObj.GetObjectKind().GroupVersionKind())
+	if err != nil {
+		return errors.Wrap(err, "getting group version resource from obj")
+	}
+	ns, err := resolveNamespace(namespace)
+	if err != nil {
+		return errors.Wrap(err, "resolving namespace")
+	}
+	if _, err := client.Resource(gvr).Namespace(ns).Patch(name, types.StrategicMergePatchType, p); err != nil {
+		return errors.Wrapf(err, "patching resource %s/%s", namespace, name)
+	}
+
+	return nil
+}
+
+func resolveNamespace(ns string) (string, error) {
 	if ns != "" {
-		return ns
+		return ns, nil
 	}
-	if m.Namespace != "" {
-		return m.Namespace
+	cfg, err := kubectx.CurrentConfig()
+	if err != nil {
+		return "", errors.Wrap(err, "getting kubeconfig")
 	}
-	return "default"
+
+	current, present := cfg.Contexts[cfg.CurrentContext]
+	if present && current.Namespace != "" {
+		return current.Namespace, nil
+	}
+	return "default", nil
 }
 
-// TODO(nkubala): change this to use the client-go dynamic client or something equally clean
-func updateRuntimeObject(client clientgo.Interface, labels map[string]string, res Artifact) error {
-	for k, v := range constants.Labels.DefaultLabels {
-		labels[k] = v
+func groupVersionResource(disco discovery.DiscoveryInterface, gvk schema.GroupVersionKind) (schema.GroupVersionResource, error) {
+	resources, err := disco.ServerResourcesForGroupVersion(gvk.GroupVersion().String())
+	if err != nil {
+		return schema.GroupVersionResource{}, errors.Wrap(err, "getting server resources for group version")
 	}
-	var err error
-	applied := false
-	var metadata *metav1.ObjectMeta
-	originalJSON, _ := json.Marshal(*res.Obj)
-	modifiedObj := (*res.Obj).DeepCopyObject()
-	for typeStr, m := range objectMetas {
-		if metadata, applied = m(modifiedObj); applied {
-			addSkaffoldLabels(labels, metadata)
-			modifiedJSON, _ := json.Marshal(modifiedObj)
-			p, _ := patch.CreateTwoWayMergePatch(originalJSON, modifiedJSON, modifiedObj)
-			err = patchers[typeStr](client, retrieveNamespace(res.Namespace, *metadata), metadata.GetName(), p)
-			break // we should only ever apply one patch, so stop here
+
+	for _, r := range resources.APIResources {
+		if r.Kind == gvk.Kind {
```

Comment:
when I hacked on this I had trouble converting the kind string into something that could be treated as a resource, mostly with capitalization (i.e. `pod` needs to be converted to `Pod`, etc.). not sure if this is something you ran into here, but might be worth addressing

Reply:
Hm, I didn't have trouble with that. The only thing that seemed a little odd was that the dynamic client needed a non-empty namespace and parsing the kubeconfig uses "" instead of "default"